src/esp/gfx/CubeMap.cpp

```diff
@@ -0,0 +1,497 @@
+
+  // color texture
+  if (flags_ & Flag::ColorTexture) {
+    auto& colorTexture = textures_[TextureType::Color];
+    colorTexture = std::make_unique<Mn::GL::CubeMapTexture>();
+    (*colorTexture)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setMinificationFilter(Mn::GL::SamplerFilter::Linear,
+                               Mn::GL::SamplerMipmap::Linear)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Linear);
+
+    if (flags_ & Flag::BuildMipmap) {
+      // RGBA8 is for the LDR. Use RGBA16F for the HDR (TODO)
+      (*colorTexture)
+          .setStorage(Mn::Math::log2(imageSize_) + 1,
+                      Mn::GL::TextureFormat::RGBA8, size);
+    } else {
+      (*colorTexture).setStorage(1, Mn::GL::TextureFormat::RGBA8, size);
+    }
+  }
+
+  // depth texture
+  if (flags_ & Flag::DepthTexture) {
+    auto& depthTexture = textures_[TextureType::Depth];
+    depthTexture = std::make_unique<Mn::GL::CubeMapTexture>();
+    (*depthTexture)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setMinificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setStorage(1, Mn::GL::TextureFormat::DepthComponent32F, size);
+  }
+}
+
+void CubeMap::recreateFramebuffer() {
+  Mn::Vector2i viewportSize{imageSize_, imageSize_};
+  for (int iFbo = 0; iFbo < 6; ++iFbo) {
+    frameBuffer_[iFbo] = Mn::GL::Framebuffer{{{}, viewportSize}};
+  }
+
+  // optional depth buffer is 24-bit integer pixel, which is different from the
+  // depth texture (32-bit float)
+  if (!(flags_ & CubeMap::Flag::DepthTexture)) {
+    for (int index = 0; index < 6; ++index) {
+      optionalDepthBuffer_[index].setStorage(
+          Mn::GL::RenderbufferFormat::DepthComponent24, viewportSize);
+    }
+  }
+}
+
+void CubeMap::attachFramebufferRenderbuffer() {
+  for (unsigned int index = 0; index < 6; ++index) {
+    if (flags_ & Flag::ColorTexture) {
+      Magnum::GL::CubeMapCoordinate cubeMapCoord =
+          convertFaceIndexToCubeMapCoordinate(index);
+      frameBuffer_[index].attachCubeMapTexture(
+          Mn::GL::Framebuffer::ColorAttachment{0},
+          *textures_[TextureType::Color], cubeMapCoord, 0);
+    }
+
+    if (flags_ & Flag::DepthTexture) {
+      Magnum::GL::CubeMapCoordinate cubeMapCoord =
+          convertFaceIndexToCubeMapCoordinate(index);
+      frameBuffer_[index].attachCubeMapTexture(
+          Mn::GL::Framebuffer::BufferAttachment::Depth,
+          *textures_[TextureType::Depth], cubeMapCoord, 0);
+    } else {
+      frameBuffer_[index].attachRenderbuffer(
+          Mn::GL::Framebuffer::BufferAttachment::Depth,
+          optionalDepthBuffer_[index]);
+    }
+  }
+}
+
+void CubeMap::prepareToDraw(unsigned int cubeSideIndex) {
+  // Note: we ONLY need to map shader output to color attachment when necessary,
+  // which means in depth texture mode, we do NOT need to do this
+  if (flags_ & CubeMap::Flag::ColorTexture) {
```

Comment:
@mosra: 
I also did tests in "DEPTH texture mode" where color output is NOT needed.
My settings:
**Control group:** has normal color buffer output.
**Test group A:** has normal color buffer output, but DISABLE the color output (magnum wrapper glColorMask(false ...) when entering rendering, and reenable it after rendering.
**Test group B:** NO color attachment, thus NO mapForDraw

Result:
Control group:
![control](https://user-images.githubusercontent.com/931093/108162724-8bf47a00-70a2-11eb-882f-8cdf017e0182.jpeg)
Test A:
![testa](https://user-images.githubusercontent.com/931093/108162734-9151c480-70a2-11eb-9ed9-d4829bf6d4d3.jpeg)
Test B:
![testB](https://user-images.githubusercontent.com/931093/108162759-9adb2c80-70a2-11eb-8bc7-ab2f9df5fc80.jpeg)

In terms of rendering time: 
Control > Test A > Test B (winner).

**No color attachment can beat the simply mask writing**, if my experiments are correct.
(PR is updated to use this method.)

Reply:
I don't want to be mean, but ... were the screenshots really necessary? Just writing out the few numbers would be enough, with this the signal:noise ratio is *really high*.